File: app/src/main/java/darkpoint/rubel/RubelWidget.java
package darkpoint.rubel;

import android.app.PendingIntent;
import android.appwidget.AppWidgetManager;
import android.appwidget.AppWidgetProvider;
import android.content.Context;
import android.content.Intent;
import android.widget.RemoteViews;
import android.widget.Toast;

import java.io.IOException;
import java.text.SimpleDateFormat;
import java.util.Calendar;
import java.util.Date;
import java.util.HashMap;
import java.util.Locale;

import retrofit2.Call;

/**
 * Implementation of App Widget functionality.
 */
public class RubelWidget extends AppWidgetProvider {

    static private int[] four_currency;
    private static boolean connection;
    static private HashMap<Integer, Rate> rateHashMap = new HashMap<>();
    static private SimpleDateFormat simpleDateFormat = new SimpleDateFormat("yyyy-MM-dd", Locale.getDefault());
    static private SimpleDateFormat simpleDateFormattop = new SimpleDateFormat("dd-MM-yyyy", Locale.getDefault());
    private static final String SYNC_CLICKED = "automaticWidgetSyncButtonClick";

    static void updateAppWidget(Context context, AppWidgetManager appWidgetManager,
                                int appWidgetId) {
        Intent intent_start = new Intent(context, SplashActivity.class);
        PendingIntent pIntent_start = PendingIntent.getBroadcast(context, appWidgetId, intent_start, 0);
        connection = MainActivity.hasConnection(context);
        // Construct the RemoteViews object
        RemoteViews views = new RemoteViews(context.getPackageName(), R.layout.rubel_widget);
        views.setOnClickPendingIntent(R.id.appwidget_text, pIntent_start);
        views.setOnClickPendingIntent(R.id.area_curs, pIntent_start);
        views.setTextViewText(R.id.wcur_one_name, "");
        views.setTextViewText(R.id.wcur_one_value, "");
        views.setTextViewText(R.id.wcur_two_name, "");
        views.setTextViewText(R.id.wcur_two_value, "");
        views.setTextViewText(R.id.wcur_three_name, "");
        views.setTextViewText(R.id.wcur_three_value, "");
        views.setTextViewText(R.id.wcur_four_name, "");
        views.setTextViewText(R.id.wcur_four_value, "");
        Intent intent = new Intent(context, RubelWidget.class);
        intent.setAction(AppWidgetManager.ACTION_APPWIDGET_UPDATE);
        intent.putExtra(AppWidgetManager.EXTRA_APPWIDGET_IDS,
                new int[]{appWidgetId});
        PendingIntent pIntent = PendingIntent.getBroadcast(context, appWidgetId, intent, 0);
        views.setOnClickPendingIntent(R.id.sync_button, pIntent);

        four_currency = Setting.loadSetting(context);
        if (connection) {
            Thread thread = new Thread(new Runnable() {
                public void run() {
                    getCurrenciesMain(four_currency, Calendar.getInstance().getTime());
                }
            });
            thread.start();
            try {
                while (true) {
                    Thread.sleep(300);
                    if (!thread.isAlive()) {
                        for (int i = 0; i < four_currency.length; i++) {
                            if (i == 0) {
                                views.setTextViewText(R.id.wcur_one_name, rateHashMap.get(four_currency[i]).Cur_Scale + " " + rateHashMap.get(four_currency[i]).Cur_Abbreviation);
                                views.setTextViewText(R.id.wcur_one_value, String.valueOf(rateHashMap.get(four_currency[i]).Cur_OfficialRate));
                            }
                            if (i == 1) {
                                views.setTextViewText(R.id.wcur_two_name, rateHashMap.get(four_currency[i]).Cur_Scale + " " + rateHashMap.get(four_currency[i]).Cur_Abbreviation);
                                views.setTextViewText(R.id.wcur_two_value, String.valueOf(rateHashMap.get(four_currency[i]).Cur_OfficialRate));
                            }
                            if (i == 2) {
                                views.setTextViewText(R.id.wcur_three_name, rateHashMap.get(four_currency[i]).Cur_Scale + " " + rateHashMap.get(four_currency[i]).Cur_Abbreviation);
                                views.setTextViewText(R.id.wcur_three_value, String.valueOf(rateHashMap.get(four_currency[i]).Cur_OfficialRate));
                            }
                            if (i == 3) {
                                views.setTextViewText(R.id.wcur_four_name, rateHashMap.get(four_currency[i]).Cur_Scale + " " + rateHashMap.get(four_currency[i]).Cur_Abbreviation);
                                views.setTextViewText(R.id.wcur_four_value, String.valueOf(rateHashMap.get(four_currency[i]).Cur_OfficialRate));
                            }
                        }
                        CharSequence widgetText = context.getString(R.string.appwidget_text) + " " + simpleDateFormattop.format(Calendar.getInstance().getTime());
                        views.setTextViewText(R.id.appwidget_text, widgetText);
                        Toast.makeText(context, "Обновлено", Toast.LENGTH_SHORT).show();
                        Setting.saveCacheCurrencyNames(context, rateHashMap.get(four_currency[0]).getCur_Scale() + " " + rateHashMap.get(four_currency[0]).getCur_Abbreviation(),
                                rateHashMap.get(four_currency[1]).getCur_Scale() + " " + rateHashMap.get(four_currency[1]).getCur_Abbreviation(),
                                rateHashMap.get(four_currency[2]).getCur_Scale() + " " + rateHashMap.get(four_currency[2]).getCur_Abbreviation(),
                                rateHashMap.get(four_currency[3]).getCur_Scale() + " " + rateHashMap.get(four_currency[3]).getCur_Abbreviation());
                        Setting.saveCacheCurrencyValues(context, (float)rateHashMap.get(four_currency[0]).getCur_OfficialRate(), (float)rateHashMap.get(four_currency[1]).getCur_OfficialRate(),
                                (float)rateHashMap.get(four_currency[2]).getCur_OfficialRate(), (float)rateHashMap.get(four_currency[3]).getCur_OfficialRate());
                        Setting.saveCacheDate(context, simpleDateFormattop.format(Calendar.getInstance().getTime()));
                        break;
                    }
                }

            } catch (Exception e) {
                appWidgetManager.updateAppWidget(appWidgetId, views);
                Toast.makeText(context, "Ошибка, проверьте подключение к интернету!", Toast.LENGTH_SHORT).show();
                return;
            }
        } else {
            CharSequence widgetText = context.getString(R.string.appwidget_text) + " " + Setting.loadCacheDate(context);
            views.setTextViewText(R.id.appwidget_text, widgetText);
            String[] cacheNames = Setting.loadCacheCurrencyNames(context);
            float[] cacheValues = Setting.loadCacheCurrencyValues(context);
            views.setTextViewText(R.id.wcur_one_name, cacheNames[0]);
            views.setTextViewText(R.id.wcur_one_value, String.valueOf(cacheValues[0]));
            views.setTextViewText(R.id.wcur_two_name, cacheNames[1]);
            views.setTextViewText(R.id.wcur_two_value, String.valueOf(cacheValues[1]));
            views.setTextViewText(R.id.wcur_three_name, cacheNames[2]);
            views.setTextViewText(R.id.wcur_three_value, String.valueOf(cacheValues[2]));
            views.setTextViewText(R.id.wcur_four_name, cacheNames[3]);
            views.setTextViewText(R.id.wcur_four_value, String.valueOf(cacheValues[3]));
            Toast.makeText(context, "Подключение к интернету отсутствует! Загружены данные последнего сеанса.", Toast.LENGTH_LONG).show();
        }
        // Instruct the widget manager to update the widget
        appWidgetManager.updateAppWidget(appWidgetId, views);
    }

    @Override
    public void onUpdate(Context context, AppWidgetManager appWidgetManager, int[] appWidgetIds) {
        // There may be multiple widgets active, so update all of them
        for (int appWidgetId : appWidgetIds) {

            updateAppWidget(context, appWidgetManager, appWidgetId);
        }
    }

    @Override
    public void onReceive(Context context, Intent intent) {
        super.onReceive(context, intent);
    }

    @Override
    public void onEnabled(Context context) {
        // Enter relevant functionality for when the first widget is created
    }

    @Override
    public void onDisabled(Context context) {
        // Enter relevant functionality for when the last widget is disabled
    }

    static private void getCurrenciesMain(final int[] currencies_id, final Date date) {
        final ServerAPI serverAPI = APIFactory.createApi();

        for (int aCurrencies_id : currencies_id) {

            Call<Rate> currencyCall = serverAPI.getRateOnDate(aCurrencies_id, simpleDateFormat.format(date));

            try {
                Rate rate = currencyCall.execute().body();
                rateHashMap.put(aCurrencies_id, rate);
            } catch (IOException e) {
                e.printStackTrace();
            }
        }
    }

}

File: app/src/main/java/darkpoint/rubel/RubelWidget.kt
package darkpoint.rubel

import android.annotation.SuppressLint
import android.app.PendingIntent
import android.appwidget.AppWidgetManager
import android.appwidget.AppWidgetProvider
import android.content.Context
import android.content.Intent
import android.graphics.Color
import android.view.View
import android.widget.RemoteViews
import android.widget.Toast
import androidx.lifecycle.MutableLiveData
import darkpoint.rubel.activities.LoadingStatus
import darkpoint.rubel.database.RubelDatabase
import darkpoint.rubel.model.Rate
import darkpoint.rubel.network.APIFactory
import kotlinx.coroutines.*
import java.text.DateFormat
import java.text.SimpleDateFormat
import java.util.*

private const val SYNCCLICKED = "syncButtonClick"

class RubelWidget : AppWidgetProvider() {

    override fun onUpdate(context: Context, appWidgetManager: AppWidgetManager, appWidgetIds: IntArray) {
        // There may be multiple widgets active, so insert all of them
        for (appWidgetId in appWidgetIds) {
            updateAppWidget(context, appWidgetManager, appWidgetId)
        }
    }

    override fun onEnabled(context: Context) {
    }

    override fun onDisabled(context: Context) {
    }

    override fun onReceive(context: Context?, intent: Intent?) {
        super.onReceive(context, intent)

        if (SYNCCLICKED == intent!!.action) {
            if (!isAlive) {
                val appmngr = AppWidgetManager.getInstance(context!!)
                val id = intent.extras.getInt("key")
                updateAppWidget(context, appmngr, id)
            } else {
                val ids = Setting.loadSetting(context!!)
                getCurrenciesMain(ids, Calendar.getInstance().time)
            }
        }
    }

    companion object {

        private lateinit var dataBase: RubelDatabase
        private var isAlive = false
        private var fourCurrency: IntArray? = null
        @SuppressLint("ConstantLocale")
        private val simpleDateFormat = SimpleDateFormat("yyyy-MM-dd", Locale.getDefault())
        private var job = Job()
        private val coroutineScope = CoroutineScope(job + Dispatchers.Main)
        private val curOne = MutableLiveData<Rate>()
        private val curTwo = MutableLiveData<Rate>()
        private val curThree = MutableLiveData<Rate>()
        private val curFour = MutableLiveData<Rate>()
        private val status = MutableLiveData<LoadingStatus>()

        internal fun updateAppWidget(context: Context, appWidgetManager: AppWidgetManager,
                                     appWidgetId: Int) {

            dataBase = RubelDatabase.getInstance(context)
            fourCurrency = Setting.loadSetting(context)
            val views = RemoteViews(context.packageName, R.layout.rubel_widget)
            views.setOnClickPendingIntent(R.id.sync_button, buildButtonPendingIntent(context, appWidgetId))

            status.observeForever {
                when (it) {
                    LoadingStatus.LOADING -> {
                        views.setViewVisibility(R.id.progressBarWidget, View.VISIBLE)
                        views.setViewVisibility(R.id.sync_button, View.GONE)
                        appWidgetManager.updateAppWidget(appWidgetId, views)
                    }
                    LoadingStatus.DONE -> {
                        Toast.makeText(context, "Обновлено", Toast.LENGTH_SHORT).show()
                        views.setTextViewText(R.id.appwidget_text, context.getString(R.string.appwidget_text, DateFormat.getDateInstance(DateFormat.LONG).format(Calendar.getInstance().time)))
                        views.setViewVisibility(R.id.progressBarWidget, View.GONE)
                        views.setViewVisibility(R.id.sync_button, View.VISIBLE)
                        appWidgetManager.updateAppWidget(appWidgetId, views)
                    }
                    else -> {
                        views.setViewVisibility(R.id.progressBarWidget, View.GONE)
                        views.setViewVisibility(R.id.sync_button, View.VISIBLE)
                        coroutineScope.launch {
                            val list = getRates()
                            if(list[0] == null){
                                Toast.makeText(context, "Ошибка! Проверьте подключение к интернету!", Toast.LENGTH_SHORT).show()
                                views.setTextViewText(R.id.appwidget_text, "Нет данных вообще!")
                                appWidgetManager.updateAppWidget(appWidgetId, views)
                                return@launch
                            }
                            curOne.value = list[0]
                            curTwo.value = list[1]
                            curThree.value = list[2]
                            curFour.value = list[3]
                            Toast.makeText(context, "Загружены данные последнего сеанса", Toast.LENGTH_SHORT).show()
                            views.setTextViewText(R.id.appwidget_text, context.getString(R.string.appwidget_text, DateFormat.getDateInstance(DateFormat.LONG).format(simpleDateFormat.parse(list[0]!!.Date))))
                            appWidgetManager.updateAppWidget(appWidgetId, views)
                        }
                        appWidgetManager.updateAppWidget(appWidgetId, views)
                    }
                }
            }

            curOne.observeForever {
                views.setTextViewText(R.id.wcur_one_name, context.getString(R.string.setting_view, it.Cur_Scale, it.Cur_Abbreviation))
                views.setTextViewText(R.id.wcur_one_value, it.Cur_OfficialRate.toString())
                views.setTextViewText(R.id.wcur_one_move, it.difference)
                views.setTextColor(R.id.wcur_one_move, it.color)
            }

            curTwo.observeForever {
                views.setTextViewText(R.id.wcur_two_name, context.getString(R.string.setting_view, it.Cur_Scale, it.Cur_Abbreviation))
                views.setTextViewText(R.id.wcur_two_value, it.Cur_OfficialRate.toString())
                views.setTextViewText(R.id.wcur_two_move, it.difference)
                views.setTextColor(R.id.wcur_two_move, it.color)
            }

            curThree.observeForever {
                views.setTextViewText(R.id.wcur_three_name, context.getString(R.string.setting_view, it.Cur_Scale, it.Cur_Abbreviation))
                views.setTextViewText(R.id.wcur_three_value, it.Cur_OfficialRate.toString())
                views.setTextViewText(R.id.wcur_three_move, it.difference)
                views.setTextColor(R.id.wcur_three_move, it.color)
            }

            curFour.observeForever {
                views.setTextViewText(R.id.wcur_four_name, context.getString(R.string.setting_view, it.Cur_Scale, it.Cur_Abbreviation))
                views.setTextViewText(R.id.wcur_four_value, it.Cur_OfficialRate.toString())
                views.setTextViewText(R.id.wcur_four_move, it.difference)
                views.setTextColor(R.id.wcur_four_move, it.color)
            }

            getCurrenciesMain(fourCurrency!!, Calendar.getInstance().time)
            appWidgetManager.updateAppWidget(appWidgetId, views)
            isAlive = true
        }

        private fun getCurrenciesMain(currencies_id: IntArray, date: Date) {
            val dateString = simpleDateFormat.format(date)
            val calendar = Calendar.getInstance()
            calendar.add(Calendar.DATE, -1)
            val dateStringY = simpleDateFormat.format(calendar.time)

            coroutineScope.launch {
                val getRateOne = APIFactory.createApi().getRateOnDateAsync(currencies_id[0], dateString)
                val getRateOneY = APIFactory.createApi().getRateOnDateAsync(currencies_id[0], dateStringY)
                val getRateTwo = APIFactory.createApi().getRateOnDateAsync(currencies_id[1], dateString)
                val getRateTwoY = APIFactory.createApi().getRateOnDateAsync(currencies_id[1], dateStringY)
                val getRateThree = APIFactory.createApi().getRateOnDateAsync(currencies_id[2], dateString)
                val getRateThreeY = APIFactory.createApi().getRateOnDateAsync(currencies_id[2], dateStringY)
                val getRateFour = APIFactory.createApi().getRateOnDateAsync(currencies_id[3], dateString)
                val getRateFourY = APIFactory.createApi().getRateOnDateAsync(currencies_id[3], dateStringY)

                try {
                    status.value = LoadingStatus.LOADING
                    val r1Y = getRateOneY.await()
                    val r1 = getRateOne.await()
                    val r2Y = getRateTwoY.await()
                    val r2 = getRateTwo.await()
                    val r3Y = getRateThreeY.await()
                    val r3 = getRateThree.await()
                    val r4Y = getRateFourY.await()
                    val r4 = getRateFour.await()
                    r1.difference = getMove(r1.Cur_OfficialRate - r1Y.Cur_OfficialRate)
                    r1.color = getMoveColor(r1.Cur_OfficialRate - r1Y.Cur_OfficialRate)
                    curOne.value = r1
                    r2.difference = getMove(r2.Cur_OfficialRate - r2Y.Cur_OfficialRate)
                    r2.color = getMoveColor(r2.Cur_OfficialRate - r2Y.Cur_OfficialRate)
                    curTwo.value = r2
                    r3.difference = getMove(r3.Cur_OfficialRate - r3Y.Cur_OfficialRate)
                    r3.color = getMoveColor(r3.Cur_OfficialRate - r3Y.Cur_OfficialRate)
                    curThree.value = r3
                    r4.difference = getMove(r4.Cur_OfficialRate - r4Y.Cur_OfficialRate)
                    r4.color = getMoveColor(r4.Cur_OfficialRate - r4Y.Cur_OfficialRate)
                    curFour.value = r4
                    insert(r1)
                    insert(r2)
                    insert(r3)
                    insert(r4)
                    status.value = LoadingStatus.DONE
                } catch (e: Exception) {
                    status.value = LoadingStatus.ERROR
                }
            }
        }

        private suspend fun getRates() : List<Rate?>{
            return withContext(Dispatchers.IO) {
                val list = listOf(
                dataBase.rubelDatabaseDao.get(fourCurrency!![0]),
                dataBase.rubelDatabaseDao.get(fourCurrency!![1]),
                dataBase.rubelDatabaseDao.get(fourCurrency!![2]),
                dataBase.rubelDatabaseDao.get(fourCurrency!![3]))
                list
            }
        }

        private suspend fun insert(rate: Rate) {
            withContext(Dispatchers.IO) {
                if(dataBase.rubelDatabaseDao.get(rate.Cur_ID) == null)
                    dataBase.rubelDatabaseDao.insert(rate)
                else
                    dataBase.rubelDatabaseDao.update(rate)
            }
        }

        private fun getMove(value: Double): String {
            return when {
                value > 0.0 -> "\uf106"
                value < 0.0 -> "\uf107"
                value == 0.0 -> "-"
                else -> ""
            }
        }

        private fun getMoveColor(value: Double): Int {
            return when {
                value > 0.0 -> Color.GREEN
                value < 0.0 -> Color.RED
                value == 0.0 -> Color.WHITE
                else -> 0
            }
        }

        private fun buildButtonPendingIntent(context: Context, key: Int): PendingIntent {
            val intent = Intent(context, RubelWidget::class.java)
            intent.action = SYNCCLICKED
            intent.putExtra("key", key)
            return PendingIntent.getBroadcast(context, 0, intent, PendingIntent.FLAG_UPDATE_CURRENT)
        }
    }
}

